File: src/main/java/io/github/numichi/reactive/logger/java/MDCContext.java
package io.github.numichi.reactive.logger.java;

import io.github.numichi.reactive.logger.exception.InvalidContextDataException;
import io.github.numichi.reactive.logger.MDC;
import io.github.numichi.reactive.logger.Values;
import reactor.core.publisher.Mono;
import reactor.util.annotation.NonNull;
import reactor.util.context.Context;
import reactor.util.context.ContextView;

import java.util.Map;
import java.util.Objects;

import static io.github.numichi.reactive.logger.exception.Messages.CTXK_NOT_NULL;
import static io.github.numichi.reactive.logger.exception.Messages.CTXW_NOT_NULL;
import static io.github.numichi.reactive.logger.exception.Messages.CTX_NOT_NULL;
import static io.github.numichi.reactive.logger.exception.Messages.MAP_NOT_NULL;
import static io.github.numichi.reactive.logger.exception.Messages.MDC_NOT_NULL;

public final class MDCContext {
    private MDCContext() {
    }
    
    @NonNull
    public static Context put(Context context, Map<String, String> mdc) {
        try {
            Objects.requireNonNull(context, CTXK_NOT_NULL);
            Objects.requireNonNull(mdc, MDC_NOT_NULL);
        } catch (NullPointerException exception) {
            throw new IllegalArgumentException(exception);
        }
        
        return context.put(Values.DEFAULT_REACTOR_CONTEXT_MDC_KEY, mdc);
    }
    
    @NonNull
    public static Context put(Context context, String mdcContextKey, Map<String, String> mdc) {
        try {
            Objects.requireNonNull(context, CTX_NOT_NULL);
            Objects.requireNonNull(mdcContextKey, CTXK_NOT_NULL);
            Objects.requireNonNull(mdc, MAP_NOT_NULL);
        } catch (NullPointerException exception) {
            throw new IllegalArgumentException(exception);
        }
        
        return context.put(mdcContextKey, mdc);
    }
    
    @NonNull
    public static Context put(Context context, MDC mdc) {
        try {
            Objects.requireNonNull(context, CTXK_NOT_NULL);
            Objects.requireNonNull(mdc, MDC_NOT_NULL);
        } catch (NullPointerException exception) {
            throw new IllegalArgumentException(exception);
        }
        
        return context.put(mdc.getContextKey(), mdc.asMap());
    }
    
    @NonNull
    public static Mono<MDC> read() {
        return read(Values.DEFAULT_REACTOR_CONTEXT_MDC_KEY);
    }
    
    @NonNull
    public static Mono<MDC> read(String mdcContextKey) {
        return Mono.deferContextual(ctx -> read(ctx, mdcContextKey));
    }
    
    @NonNull
    public static Mono<MDC> read(ContextView context) {
        return read(context, Values.DEFAULT_REACTOR_CONTEXT_MDC_KEY);
    }
    
    @NonNull
    public static Mono<MDC> read(@NonNull ContextView contextView, @NonNull String mdcContextKey) {
        try {
            Objects.requireNonNull(contextView, CTXW_NOT_NULL);
            Objects.requireNonNull(mdcContextKey, CTXK_NOT_NULL);
        } catch (NullPointerException exception) {
            return Mono.error(new IllegalArgumentException(exception));
        }
        
        MDC mdc = new MDC(mdcContextKey);
        
        try {
            Map<String, String> map = contextView.get(mdcContextKey);
            mdc.putAll(map);
        } catch (Exception exception) {
            return Mono.error(new InvalidContextDataException(exception));
        }
        
        return Mono.just(mdc);
    }
}

File: kotlin-coroutine-reactive-logger/src/main/kotlin/hu/numichi/reactive/logger/kotlin/MDCContext.kt
package hu.numichi.reactive.logger.kotlin;

import hu.numichi.reactive.logger.Consts.DEFAULT_REACTOR_CONTEXT_MDC_KEY
import hu.numichi.reactive.logger.MDC
import hu.numichi.reactive.logger.exception.InvalidContextDataException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.reactor.asCoroutineContext
import reactor.util.context.Context
import reactor.util.context.ContextView

object MDCContext {

    suspend fun read(): MDC {
        return read(DEFAULT_REACTOR_CONTEXT_MDC_KEY)
    }

    suspend fun read(mdcContextKey: String): MDC {
        return read(rectorContext(), mdcContextKey)
    }

    fun read(context: ContextView?): MDC {
        return read(context, DEFAULT_REACTOR_CONTEXT_MDC_KEY)
    }

    fun read(contextView: ContextView?, mdcContextKey: String): MDC {
        if (contextView == null) {
            throw NullPointerException("ContextView must not be null or this context is not exist")
        }

        val mdc = MDC(mdcContextKey)

        try {
            val map: Map<String, String> = contextView.get(mdcContextKey)
            mdc.putAll(map)
        } catch (exception: Exception) {
            throw InvalidContextDataException(exception)
        }

        return mdc
    }

    suspend fun <T> withContext(vararg mdc: MDC, block: suspend CoroutineScope.() -> T): T {
        return withContext(mdc.toList(), rectorContext(), block)
    }

    suspend fun <T> withContext(context: Context, vararg mdc: MDC, block: suspend CoroutineScope.() -> T): T {
        return withContext(mdc.toList(), context, block)
    }

    private suspend fun <T> withContext(mdcCollection: Collection<MDC>, context: Context?, block: suspend CoroutineScope.() -> T): T {
        var newContext = context ?: Context.empty()

        mdcCollection.forEach {
            newContext = newContext.put(it.contextKey, it.map)
        }

        return kotlinx.coroutines.withContext(newContext.asCoroutineContext()) {
            block()
        }
    }
}
